pyqir-generator/pyqir/generator/_native.pyi

```diff
@@ -29,30 +29,55 @@ def bitcode_to_ir(bitcode: bytes, module_name: Optional[str], source_file_name:
     ...
 
 
-class Qubit:
-    """A qubit identifier."""
-    ...
-
-
 class ResultRef:
     """A mutable reference cell that holds a measurement result."""
     ...
 
 
 class Function:
-    """A callable value for a module function."""
+    """A QIR function."""
     ...
 
 
+class Value:
```

Comment:
A point that is slightly confounding to me is the multiple definition of the `Value` type. For instance, here, it looks like a module-wide definition for the purpose of holding either an integer or a double type for (not only) qubit indexing and gate parametrisation typing. This is to be contrasted with the definition in `types.py` (see [here](https://github.com/qir-alliance/pyqir/blob/38dd3d0540a646588b1ed3cfc053de3097581dc8/pyqir-generator/pyqir/generator/types.py#L42)) which looks like the type exposed to the user API.

Reply:
- `pyqir.generator._native.Value` is a QIR/LLVM value.
- `pyqir.generator.types.Value` is a QIR/LLVM type, restricted to the set of types that a value can be (that is, excluding void).

I agree that the naming is confusing, though. I think we don't actually need to keep the void type separate, so I combined `types.Value` and `types.Return` into one thing called `Type`.

Reply:
Thanks for the explanation. At the risk of being a bit pedantic, it looks to me that this is redundant now with the use of types in the `types` module with the exception of `BOOL`. Just a suggestion: would it make sense to keep the distinction and name them `ValueType` and `ReturnType` ?

Reply:
Sorry, I'm not sure what you mean by `Value` being redundant, could you explain?

Reply:
Ah, sorry I've been unclear. I meant the fact that QIR/LLVM types are defined [here](https://github.com/qir-alliance/pyqir/blob/ed8c20b2a216a7e0e79a49964b24e7692bf4e191/pyqir-generator/pyqir/generator/types.py#L53) but also accessed through the aliases (`types.<type>`). In short, I am wondering if the definition of aliases is necessary. But there might be a need that eludes me now.

Reply:
Do you mean these aliases?

```python
VOID: Void = Void._VOID
DOUBLE: Double = Double._DOUBLE
QUBIT: Qubit = Qubit._QUBIT
RESULT: Result = Result._RESULT
```

That's an artifact of how I'm using `Union` to define a sum type - the nullary constructors are awkward. Making them singleton types seemed like the least worst option, but I didn't want to make people write `Double.DOUBLE`, so I defined module-level aliases for the singleton instances.

Reply:
Ah OK. I understand now. This is because when `types.Function` is instantiated, it takes as argument a list of references to types instances. Thanks.

Reply:
The issue exists even if the type is not recursive. For example, how can this Rust type be encoded in Python:

```rust
enum Sum {
    Foo,
    Bar { value: i64 },
}
```

As a union, each case is its own type and could be encoded like:

```python
@dataclass
class Foo: pass

@dataclass
class Bar:
    value: int

Sum = Foo | Bar
```

But now you have to construct and pattern match on `Foo()`, not `Foo` (the parentheses are annoying), and each time you construct `Foo()` a new object in memory is instantiated (which is unnecessary). So I make `Foo` a singleton using `Enum`:

```python
class Foo(Enum):
    FOO = auto()
```

but still `Foo.FOO` is the member of `Sum`, not `Foo` itself. So I add the module-level alias to `Foo`'s instance for convenience.

Reply:
Yep, I get that. I was just wondering if doing something as simple as:

```python
class Sum(Enum):
     Foo
     Bar: int
```
Wouldn't allow to do what you want.

Reply:
In Python, enum cases are like static fields, not constructors. So every case needs to be of the form `name = value`. `Sum.Bar` isn't expressible as an enum case because you can't construct `Sum.Bar(x)` for an arbitrary *x*. In the definition you gave, `Foo` and `Bar` are uninitialized attributes and can't be used.